File: src/queries/geo.ts
import { rawQuery, buildFilterSQL, type Filter } from "@/lib/queries";

export interface GeoMetrics {
  country: string;
  region: string | null;
  city: string | null;
  visitors: number;
  pageviews: number;
  visits: number;
  bounceRate: number;
  avgDuration: number;
}

/**
 * Get geographic breakdown with bounce rate and avg duration.
 * Demographic fields (country/region/city) live on the session table.
 */
export async function getGeoMetrics(
  websiteId: string,
  startDate: Date,
  endDate: Date,
  level: "country" | "region" | "city" = "country",
  countryFilter?: string,
  regionFilter?: string,
  limit = 50,
  filters: Filter[] = []
): Promise<GeoMetrics[]> {
  let groupBy: string;
  let selectGeo: string;
  let extraFilter = "";
  const params: Record<string, unknown> = { websiteId, startDate, endDate };

  // Merge filter params
  const { sql: filterWhere, params: filterParams } = buildFilterSQL(filters);
  Object.assign(params, filterParams);

  switch (level) {
    case "city":
      groupBy = "s.country, s.region, s.city";
      selectGeo = "s.country, s.region, s.city,";
      if (countryFilter) {
        extraFilter += " AND s.country = {{countryFilter}}";
        params.countryFilter = countryFilter;
      }
      if (regionFilter) {
        extraFilter += " AND s.region = {{regionFilter}}";
        params.regionFilter = regionFilter;
      }
      break;
    case "region":
      groupBy = "s.country, s.region";
      selectGeo = "s.country, s.region, NULL AS city,";
      if (countryFilter) {
        extraFilter += " AND s.country = {{countryFilter}}";
        params.countryFilter = countryFilter;
      }
      break;
    default:
      groupBy = "s.country";
      selectGeo = "s.country, NULL AS region, NULL AS city,";
  }

  const results = await rawQuery<{
    country: string;
    region: string | null;
    city: string | null;
    visitors: bigint;
    pageviews: bigint;
    visits: bigint;
    bounce_rate: number;
    avg_duration: number;
  }>(
    `WITH visit_stats AS (
      SELECT
        ${groupBy},
        we.visit_id,
        COUNT(*) AS pages,
        EXTRACT(EPOCH FROM (MAX(we.created_at) - MIN(we.created_at))) AS duration
      FROM website_event we
      JOIN session s ON s.session_id = we.session_id
      WHERE we.website_id = {{websiteId::uuid}}
        AND we.created_at BETWEEN {{startDate::timestamptz}} AND {{endDate::timestamptz}}
        AND we.event_type = 1
        AND s.country IS NOT NULL
        ${extraFilter}
        ${filterWhere}
      GROUP BY ${groupBy}, we.visit_id
    )
    SELECT
      ${selectGeo}
      COUNT(DISTINCT visit_id)::bigint AS visitors,
      SUM(pages)::bigint AS pageviews,
      COUNT(visit_id)::bigint AS visits,
      CASE WHEN COUNT(*) = 0 THEN 0
        ELSE (SUM(CASE WHEN pages = 1 THEN 1 ELSE 0 END)::float / COUNT(*) * 100)
      END AS bounce_rate,
      COALESCE(AVG(duration), 0) AS avg_duration
    FROM visit_stats
    GROUP BY ${groupBy}
    ORDER BY visitors DESC
    LIMIT ${limit}`,
    params
  );

  return results.map((r) => ({
    country: r.country,
    region: r.region,
    city: r.city,
    visitors: Number(r.visitors),
    pageviews: Number(r.pageviews),
    visits: Number(r.visits),
    bounceRate: r.bounce_rate ?? 0,
    avgDuration: r.avg_duration ?? 0,
  }));
}

File: core/mantecato_core/queries/geo.py
"""Geographic breakdown queries — aggregate pageviews by country.

Privacy-first: only country-level aggregation. Region/city removed to
prevent re-identification of individuals in low-traffic scenarios.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from datetime import datetime

from core.mantecato_core.database import raw_query
from core.mantecato_core.filters import Filter, prepare_filters
from core.mantecato_core.queries.orm_fallbacks import (
    count_by_field,
    pageview_queryset,
    should_use_orm_fallback,
)


def get_geo_metrics(
    website_id: str,
    start_date: datetime,
    end_date: datetime,
    limit: int = 50,
    filters: list[Filter] | None = None,
) -> list[dict[str, Any]]:
    """Aggregate pageview counts by country (ISO 3166-1 alpha-2)."""
    if should_use_orm_fallback():
        rows = count_by_field(
            pageview_queryset(website_id, start_date, end_date, filters),
            "country",
            "pageviews",
            limit,
        )
        total = sum(int(row["pageviews"] or 0) for row in rows)
        return [
            {
                "country": row["value"],
                "pageviews": int(row["pageviews"] or 0),
                "percentage": round((int(row["pageviews"] or 0) / total) * 100, 1)
                if total > 0
                else 0,
            }
            for row in rows
        ]

    filters = filters or []
    filter_where, filter_params, _ = prepare_filters(filters)

    params: dict[str, Any] = {
        "websiteId": website_id,
        "startDate": start_date,
        "endDate": end_date,
        **filter_params,
    }

    rows = raw_query(
        """SELECT
      we.country AS country,
      COUNT(*)::bigint AS pageviews
    FROM website_event we
    WHERE we.website_id = {websiteId::uuid}
      AND we.created_at BETWEEN {startDate::timestamptz} AND {endDate::timestamptz}
      AND we.event_type = 1
      AND we.country IS NOT NULL
      AND we.country != ''
      """ + filter_where + """
    GROUP BY we.country
    ORDER BY pageviews DESC
    LIMIT """ + str(limit),
        params,
    )

    total = sum(int(r["pageviews"] or 0) for r in rows)

    return [
        {
            "country": row["country"],
            "pageviews": int(row["pageviews"] or 0),
            "percentage": round((int(row["pageviews"] or 0) / total) * 100, 1) if total > 0 else 0,
        }
        for row in rows
    ]
